src/frontend/hooks/useMapStyles.ts

```
@@ -0,0 +1,113 @@
+    backgroundMaps ? "loading" : "success"
+  );
+
+  useFocusEffect(
+    React.useCallback(() => {
+      if (!backgroundMaps) return;
+      let didCancel = false;
+      api.maps
+        .getStyleList()
+        .then(styles => {
+          if (didCancel) return;
+          setStylesList(styles);
+          setStatus("success");
+        })
+        .catch(() => didCancel || setStatus("error"));
+
+      return () => {
+        didCancel = true;
+      };
+    }, [backgroundMaps])
+  );
+
+  // Keeping this in-scope so we can add translated name later
+  const defaultStyle = React.useMemo(() => {
+    return {
+      id: DEFAULT_MAP_ID,
+      url: onlineStyleURL,
+      bytesStored: 0,
+      name: "Default",
+    };
+  }, []);
+
+  const customStyle: MapServerStyleInfo = React.useMemo(() => {
+    return {
+      id: CUSTOM_MAP_ID,
+      url: api.getMapStyleUrl("default"),
+      bytesStored: 0,
+      name: "Offline Map",
+    };
+  }, []);
+
+  const mergedStatus: Status =
+    customStyleAvailability === "unknown" || status === "loading"
+      ? "loading"
+      : status;
+
+  let styles: StylesAtLeastOne;
+
+  if (backgroundMaps) {
+    // Always show the online default style in MapServer. This is replaced with
+    // fallback in the useMapStyle hook
+    styles =
+      customStyleAvailability === "available"
+        ? [defaultStyle, customStyle, ...stylesList]
+        : [defaultStyle, ...stylesList];
+  } else {
+    if (customStyleAvailability === "available") {
+      styles = [customStyle];
+    } else {
+      styles = [defaultStyle];
+    }
+  }
+
+  const setSelectedStyleId = React.useCallback(
+    (styleId: string) => {
+      setSettings("mapStyleId", styleId);
+    },
+    [setSettings]
+  );
+
+  let selectedStyleId = settings.mapStyleId;
+  // If the user selected style is not available, default to the first in the
+  // styles list.
+  if (!styles.find(style => style.id === selectedStyleId)) {
+    selectedStyleId = styles[0].id;
+  }
```

We should also probably throw an error, otherwise it could look like it is just not working

---

l think it makes sense to keep this as is because the first item in `styles` is the default/fallback style and consumers of this hook may still be waiting for updates to that styles state from the server for some period of time, so we wouldn't just want to prematurely throw an error. that's what i gather from this, at least.

---

The typing for `styles` ensures that it always contains at least one item in the array. You can rely on the map server either being in a "loading" state, or this to be loaded with at least one style available.